apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+                    if (!_isReading || _currentCall.ResponseStream.Current == null)
+                    {
+                        return false;
+                    }
+                    if (_metadata == null)
+                    {
+                        _metadata = _currentCall.ResponseStream.Current.Metadata;
+                    }
+                }
+                RecordResumeToken();
+            }
+            return _isReading;
+        }
+
+        private async Task<bool> ReliableMoveNext(CancellationToken cancellationToken)
+        {
+            try
+            {
+                Logger.LogPerformanceCounterFn("StreamReader.MoveNextCount", x => x + 1);
+                //we increment our skip count before calling MoveNext so that a reconnect operation
+                //will fast forward to the proper place.
+                _resumeSkipCount++;
+                _isReading = await _currentCall.ResponseStream.MoveNext(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e) 
+            {
+                _currentCall = null;
+
+                //reconnect on failure which will call reliableconnect and respect resumetoken and resumeskip
+                cancellationToken.ThrowIfCancellationRequested();
+
+                Logger.Warn(() => $"An error occurred attemping to iterate through the sql query.  Attempting to recover. Exception:{e}");
+
+                //when we reconnect, we purposely do not do a *reliable*movenext.  If we fail to fast forward on the reconnect
+                //we bail out completely and surface the error.
+                return await ReliableConnect(cancellationToken);
+            }
+            RecordResumeToken();
+            return _isReading;
+        }
+
+        private void RecordResumeToken()
+        {
+            if (_isReading && _currentCall != null)
+            {
+                //record resume information.
+                if (_currentCall.ResponseStream.Current.ResumeToken != null)
+                {
+                    _resumeToken = _currentCall.ResponseStream.Current.ResumeToken;
+                    _resumeSkipCount = 0;
+                }
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<ResultSetMetadata> GetMetadataAsync(CancellationToken cancellationToken)
+        {
+            await ReliableConnect(cancellationToken).ConfigureAwait(false);
+            return _metadata;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public bool IsClosed { get; private set; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public Session Session => _session;
+
+        /// <summary>
+        /// 
+        /// 
+        /// </summary>
+        public void Close()
+        {
+            lock (_sync)
```

Comment:
It's not clear why locking is used just for this part. It looks like this isn't really thread-safe elsewhere anyway - for example, `RecordResumeToken` checks `ResponseStream.Current.ResumeToken` for nullity, then fetches it again... if another call to `ReliableMoveNext` occurs between those checks, it could be null again (I'd guess, at least).

Is `Close` different because it's more likely to *be* called from multiple threads, and we need to make it safer?